### PR TITLE
fix silent overflows on `Step` impls

### PR DESCRIPTION
**Pull request by matthew-piziak**
Part of https://github.com/rust-lang/rust/issues/36110

r? @eddyb

### CODE REVIEW

File: src/libcoretest/iter.rs [hunk 1]
```diff
@@ -916,6 +916,20 @@ fn test_range_step() {
 }
 
 #[test]
+#[should_panic]
+fn test_range_overflow_unsigned() {
+    let mut it = u8::MAX..;
```
Comment (sfackler) [inline src/libcoretest/iter.rs]: Won't these tests fail if compiled in release mode?

Comment (eddyb) [inline src/libcoretest/iter.rs]: Ahh, that's a good point. Then these have to be run-fail with the overflow checking forced on (see `src/test/run-fail/*overflow*.rs`).

Comment (sfackler) [inline src/libcoretest/iter.rs]: See #36372 for reference on making these tests.

**Comment by bluss** [comment]
Test doesn't pass and needs update

**Comment by matthew-piziak** [comment]
Sorry for the inactivity. I've been traveling for a few weeks.

@alexcrichton Can you elaborate, or link the PR if you remember which one it was?

File: src/libcore/iter/range.rs [hunk 2]
```diff
@@ -95,11 +95,13 @@ macro_rules! step_impl_unsigned {
             }
 
             #[inline]
+            #[rustc_inherit_overflow_checks]
             fn add_one(&self) -> Self {
                 *self + 1
```
Comment (eddyb) [inline src/libcore/iter/range.rs]: `Add::add(*self, 1)` is what @alexcrichton meant. Similar situation on the rest.

**Comment by alexcrichton** [comment]
Looks like travis failures may be legit?

**Comment by matthew-piziak** [comment]
I'm missing something here. My tests are failing on explicit panic instead of overflowing as expected.

**Comment by alexcrichton** [comment]
The current failures look like:

```
failures:
thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:298

---- [run-fail] run-fail/iter-step-overflow-debug.rs stdout ----
	
error: failure produced the wrong error: exit code: 0
status: exit code: 0
command: /build/build/x86_64-unknown-linux-gnu/test/run-fail/iter-step-overflow-debug.stage2-x86_64-unknown-linux-gnu 
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
thread 'main' panicked at 'attempt to add with overflow', src/libcore/ops.rs:315
note: Run with `RUST_BACKTRACE=1` for a backtrace.
thread 'main' panicked at 'attempt to add with overflow', src/libcore/ops.rs:315

------------------------------------------

thread '[run-fail] run-fail/iter-step-overflow-debug.rs' panicked at 'explicit panic', src/tools/compiletest/src/runtest.rs:2377
note: Run with `RUST_BACKTRACE=1` for a backtrace.

---- [run-fail] run-fail/iter-step-overflow-ndebug.rs stdout ----
	
error: no error pattern specified in "/build/src/test/run-fail/iter-step-overflow-ndebug.rs"
thread '[run-fail] run-fail/iter-step-overflow-ndebug.rs' panicked at 'explicit panic', src/tools/compiletest/src/runtest.rs:1588


failures:
    [run-fail] run-fail/iter-step-overflow-debug.rs
    [run-fail] run-fail/iter-step-overflow-ndebug.rs
```

I think these tests should be `run-pass` tests instead of `run-fail` tests perhaps?

**Comment by alexcrichton** [comment]
Looks like travis may unfortunately still be failing?

```
---- [run-pass] run-pass/iter-step-overflow-ndebug.rs stdout ----
	
error: test run failed!
status: exit code: 101
command: /build/build/x86_64-unknown-linux-gnu/test/run-pass/iter-step-overflow-ndebug.stage2-x86_64-unknown-linux-gnu 
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
thread 'main' panicked at 'assertion failed: `(left == right)` (left: `255`, right: `0`)', /build/src/test/run-pass/iter-step-overflow-ndebug.rs:15
note: Run with `RUST_BACKTRACE=1` for a backtrace.

------------------------------------------

thread '[run-pass] run-pass/iter-step-overflow-ndebug.rs' panicked at 'explicit panic', src/tools/compiletest/src/runtest.rs:2377
note: Run with `RUST_BACKTRACE=1` for a backtrace.


failures:
    [run-pass] run-pass/iter-step-overflow-ndebug.rs
```

**Comment by steveklabnik** [comment]
Looks like travis is passing now, r? @eddyb

**Review by eddyb — APPROVED** [review]

**Pull request merged**